client/src/main/java/io/pravega/client/stream/impl/ReaderGroupStateManager.java

```diff
@@ -267,6 +269,15 @@ private void fetchUpdatesIfNeeded() {
             sync.fetchUpdates();
             long groupRefreshTimeMillis = sync.getState().getConfig().getGroupRefreshTimeMillis();
             fetchStateTimer.reset(Duration.ofMillis(groupRefreshTimeMillis));
+            compactIfNeeded();
+        }
+    }
+    
+    private void compactIfNeeded() {
+        ReaderGroupState state = sync.getState();
+        //Make sure it has been a while, and compaction are staggered.
+        if (state.getUpdatesSinceCompaction() > MIN_UPDATES_BEFORE_COMPACTION && Math.random() < COMPACTION_PROBABILITY) {
```

Comment:
What is the disadvantage of performing compaction after `MIN_UPDATES_BEFORE_COMPACTION` without the check for for compaction probability?

Reply:
This is introducing some randomization to the compaction to avoid having all readers performing checkpoints in a synchronized manner. I'm actually wondering whether we are resetting the counter once the reader finds a checkpoint. I don't quite follow the flow here. Does it instantiate `ReaderGroupState` every time it pulls a checkpoint from the synchronizer?